sendrecv/gst-rust/src/main.rs

```diff
@@ -0,0 +1,426 @@
+    ServerRegistering = 2000,
+    ServerRegisteringError,
+    ServerRegistered,
+    ServerClosed,
+    PeerConnecting = 3000,
+    PeerConnectionError,
+    PeerConnected,
+    PeerCallNegotiating = 4000,
+    PeerCallStarted,
+    PeerCallStopping,
+    PeerCallStopped,
+    PeerCallError,
+    None,
+}
+
+fn check_plugins() -> bool {
+    let needed = vec![
+        "opus",
+        "vpx",
+        "nice",
+        "webrtc",
+        "dtls",
+        "srtp",
+        "rtpmanager",
+        "videotestsrc",
+        "audiotestsrc",
+    ];
+    let registry = gst::Registry::get();
+    let mut ret = true;
+    for plugin_name in needed {
+        let plugin = registry.find_plugin(&plugin_name.to_string());
+        if plugin.is_none() {
+            println!("Required gstreamer plugin '{}' not found", plugin_name);
+            ret = false;
+        }
+    }
+    return ret;
+}
+
+fn setup_call(app_state: &Arc<AtomicUsize>, out: &ws::Sender) -> AppState {
+    app_state.store(AppState::PeerConnecting as usize, Ordering::Relaxed);
+    println!("Setting up signalling server call with 1");
+    out.send("SESSION 1").unwrap();
+    return AppState::PeerConnecting;
+}
+
+fn register_with_server(app_state: &Arc<AtomicUsize>, out: &ws::Sender) -> AppState {
+    app_state.store(AppState::ServerRegistering as usize, Ordering::Relaxed);
+    let our_id = rand::thread_rng().gen_range(10, 10_000);
+    println!("Registering id {} with server", our_id);
+    out.send(format!("HELLO {}", our_id)).unwrap();
+    return AppState::ServerRegistering;
+}
+
+fn sdp_message_as_text(offer: gstreamer_webrtc::WebRTCSessionDescription) -> Option<String> {
+    unsafe {
+        from_glib_full(gstreamer_sdp_sys::gst_sdp_message_as_text(
+            (*offer.to_glib_none().0).sdp,
+        ))
+    }
+}
+
+fn send_sdp_offer(
+    app_state: &Arc<AtomicUsize>,
+    offer: gstreamer_webrtc::WebRTCSessionDescription,
+    out: &ws::Sender,
+) {
+    if app_state.load(Ordering::Relaxed) < (AppState::PeerCallNegotiating as usize) {
+        // TODO signal and cleanup
+        panic!("Can't send offer, not in call");
+    };
+    let message = json!({
+          "sdp": {
+            "type": "offer",
+            "sdp": sdp_message_as_text(offer).unwrap(),
+          }
+        });
+    out.send(message.to_string()).unwrap();
+}
+
+fn on_offer_created(
+    app_state: &Arc<AtomicUsize>,
+    webrtc: gst::Element,
+    promise: &gst::Promise,
+    out: &ws::Sender,
+) {
+    assert_eq!(
+        app_state.load(Ordering::Relaxed),
+        AppState::PeerCallNegotiating as usize
+    );
+    let reply = promise.get_reply().unwrap();
+
+    let offer = reply
+        .get_value("offer")
+        .unwrap()
+        .get::<gstreamer_webrtc::WebRTCSessionDescription>()
+        .expect("Invalid argument");
+    let promise = gst::Promise::new();
+    webrtc
+        .emit(
+            "set-local-description",
+            &[&offer.to_value(), &promise.to_value()],
```

Comment:
Hmm what's the point of the promise if it's not used anywhere afterwards?

Reply:
I was mostly parroting the original c implementation: https://github.com/centricular/gstwebrtc-demos/blob/master/sendrecv/gst/webrtc-sendrecv.c#L252-L255

It does seem though that no promise is allowed: https://gstreamer.freedesktop.org/data/doc/gstreamer/head/gst-plugins-bad/html/gst-plugins-bad-plugins-webrtcbin.html#GstWebRTCBin-set-local-description

On that point, I couldn't figure out how to pass null in a situation like this. Is there some stand-in in the library?

Reply:
You should be able to pass `&None::<gst::Promise>` here